db/transactions.go

```diff
@@ -21,79 +21,85 @@ func NewTxManager(db *sqlx.DB) *TxManager {
 	return &TxManager{DB: db}
 }
 
-func (s *TxManager) WithTransaction(ctx context.Context, f func(context.Context, Querier) error) (err error) {
-	// Retry this transaction 3 times
-	for i := 0; i < 3; i++ {
-		err = s.withOneTransaction(ctx, i, f)
-		if !retryable(err) {
-			break
-		}
-	}
-	return err
-}
+type queryFn func(context.Context, Querier) error
 
-func retryable(err error) bool {
-	return errors.Is(err, ErrBadConn)
-}
+// WithTransaction wraps f in an explicit o11y'd transaction, handling rollback
+// if f returns an error.
+func (s *TxManager) WithTransaction(ctx context.Context, f queryFn) (err error) {
+	const maxAttempts = 3
 
-func (s *TxManager) withOneTransaction(ctx context.Context,
-	attempt int, f func(context.Context, Querier) error) (err error) {
-	ctx, span := o11y.StartSpan(ctx, "tx-manager: with-one-transaction")
-	defer func() {
-		if retryable(err) {
-			o11y.AddField(ctx, "bad_con", attempt)
-		}
-		o11y.End(span, &err)
-	}()
-	span.AddField("attempt", attempt)
-	tx, err := s.DB.BeginTxx(ctx, nil)
+	transaction := func(ctx context.Context, attempt int, f queryFn) (err error) {
+		ctx, span := o11y.StartSpan(ctx, "tx-manager: tx")
+		defer func() {
+			// don't report errors on bad connections unless all the retries failed
+			if errors.Is(err, ErrBadConn) {
+				o11y.AddField(ctx, "bad_con", attempt)
+				if attempt <= maxAttempts {
```

Comment:
It feels slightly better to internalise the func and `maxAttempts` so that the relationship between the different flow for `errors.Is(err, ErrBadConn)` and `maxAttempts` is clearly only used in this method.

I think the important thing is just to keep our errors actionable - so the inner transaction reporting bad_con - and not error on the span is the main goal.

If you do preferred to split this back out then the `maxAttempts` const will need to be shared so that we can still do this check.